api/src/utils/merge-version-data.ts

```diff
@@ -0,0 +1,186 @@
+import type { Item, SchemaOverview } from '@directus/types';
+import Joi from 'joi';
+import type { Alterations } from '../types/index.js';
+
+const alterationSchema = Joi.object({
+	create: Joi.array().items(Joi.object().unknown()),
+	update: Joi.array().items(Joi.object().unknown()),
+	delete: Joi.array().items(Joi.string(), Joi.number()),
+});
+
+export function mergeVersionsRaw(item: Item, versionData: Partial<Item>[]) {
+	const result = { ...item };
+
+	for (const versionRecord of versionData) {
+		for (const key in versionRecord) {
+			result[key] = versionRecord[key];
+		}
+	}
+
+	return result;
+}
+
+export function mergeVersionsRecursive(
+	item: Item,
+	versionData: Partial<Item>[],
+	collection: string,
+	schema: SchemaOverview,
+): Item {
+	if (versionData.length === 0) return item;
+
+	return recursiveMerging(item, versionData, collection, schema) as Item;
+}
+
+function recursiveMerging(
+	data: Item,
+	versionData: Partial<Item>[],
+	collection: string,
+	schema: SchemaOverview,
+): unknown {
+	const result = { ...data };
+	const relations = getRelations(collection, schema);
+
+	for (const versionRecord of versionData) {
+		for (const key in data) {
+			if (!versionRecord || key in versionRecord === false) {
+				continue;
+			}
+
+			const currentValue: unknown = data[key];
+			const newValue: unknown = versionRecord[key];
+
+			if (typeof newValue !== 'object' || newValue === null) {
+				// primitive type substitution
+				result[key] = newValue;
+				continue;
+			}
+
+			if (key in relations === false) {
+				// check for m2a exception
+				if (isManyToAnyCollection(collection, schema) && key === 'item') {
+					const item = addMissingKeys(currentValue && typeof currentValue === 'object' ? currentValue : {}, newValue);
```

Comment:
```suggestion
					const item = addMissingKeys(isObject(currentValue) ? currentValue : {}, newValue);
```

If accepted an import from utils is required for this to work.